examples/mxnet_imagenet_resnet50.py

```diff
@@ -0,0 +1,325 @@
+class LogValidationMetricsCallback(object):
+
+    def __call__(self, param):
+        if not param.eval_metric:
+            return
+        name_value = param.eval_metric.get_name_value()
+        for name, value in name_value:
+            val_array = mx.nd.array([value])
+            hvd.allreduce_(val_array, average=True, name=str(-1))
+            logging.info('Epoch[%d] Validation-%s=%f', param.epoch,
+                         name, val_array.asscalar())
+
+# Fetch training and validation data if present
+try:
+    train_data, val_data, batch_fn = get_data_rec(args.rec_train,
+                                                  args.rec_train_idx,
+                                                  args.rec_val,
+                                                  args.rec_val_idx,
+                                                  batch_size,
+                                                  data_nthreads)
+except Exception as e:
+    print("Data not found, using synthetic data instead!")
+    args.benchmark = 1
+
+# Otherwise use synthetic data
+if args.benchmark is 1:
+    image_shape = (3, 224, 224)
+    data_shape = (batch_size,) + image_shape
+    train_data = SyntheticDataIter(classes, data_shape, epoch_size, np.float32)
+    val_data = None
+
+
+def main():
+    # Get model from Gluon model zoo
+    # https://mxnet.incubator.apache.org/api/python/gluon/model_zoo.html
+    net = get_model(args.model, **kwargs)
+    net.cast(args.dtype)
+
+    # Create input symbol
+    data = mx.sym.var('data')
+    if args.dtype == 'float16':
+        data = mx.sym.Cast(data=data, dtype=np.float16)
+        net.cast(np.float16)
+
+    # Create output symbol
+    out = net(data)
+    if args.dtype == 'float16':
+        out = mx.sym.Cast(data=out, dtype=np.float32)
+    softmax = mx.sym.SoftmaxOutput(out, name='softmax')
+
+    if args.use_pretrained:
+        arg_params = {} 
+        for x in net.collect_params().values():
+            x.reset_ctx(mx.cpu())
+            arg_params[x.name] = x.data()
+    else:
+        arg_params = None
+    aux_params = None
+
+    # Create model
+    mod = mx.mod.Module(softmax, context=context)
+
+    # Create optimizer
+    optimizer_params = {'wd': args.wd,
+                        'momentum': args.momentum,
+                        'lr_scheduler': lr_scheduler,
+                        'multi_precision': True}
+    opt = mx.optimizer.create('sgd', sym=out, **optimizer_params)
+    opt = hvd.DistributedOptimizer(opt)
+
+    # Create initializer and initializer parameters
+    initializer = mx.init.Xavier(rnd_type='gaussian', factor_type="in",
+                                 magnitude=2)
+    mod.bind(data_shapes=train_data.provide_data,
+             label_shapes=train_data.provide_label)
+    mod.init_params(initializer, arg_params=arg_params, aux_params=aux_params)
+
+    # Fetch and broadcast parameters
+    (arg_params, aux_params) = mod.get_params()
+    mx.nd.waitall()
```

Comment:
Why do we need these waitalls here?  Is it because we don't have ReadyEvent in Horovod adapter?